The caching algorithm should observe the max-age headers

ashleyfrieze opened the pull request:
This is the foremost priority, dropping back to `Last-Modified` with a heuristic as per RFC7234

This replaces the behaviour where the max-age headers were not considered at the time of caching, and where very recently `Last-Modified` pages were NOT allowed in the cache either way.

This change introduces some additional tests, with some test refactoring to avoid too many repeated constants.

It also centralises the logic for eviction and ingress into the cache.

It's a potential fix for #432

ashleyfrieze committed:
The caching algorithm should observe the max-age headers first and foremost, dropping back to `Last-Modified` with a heuristic as per RFC7234

rbri commented:
Great thanks, can you please add your real name to the list of authors for both changed files.

rbri commented:
and on my machine contentWithLastModifiedTimeIsCachedAfterAFewPercentOfCreationAge fails - maybe because of date format issues (only a wild guess have not analyzed this)

ashleyfrieze commented:
@rbri test failure was a minor test design issue - mixing hard coded time with calculated times. My mistake. I was rushing on a Friday evening. I've changed it to use per test constants relating to `now` and that should sort it out.

I've also modified the authors as requested.

rbri commented:
many thanks, feel free to provide more of this
